### ISSUE TITLE
Save document through Yjs websocket

### ISSUE THREAD

**Issue by davidbrochart** [jupyterlab/jupyterlab#11702]
### Problem

We are working towards saving documents from the back-end, instead of sending the content from the front-end using the `PUT /api/contents/{path}` HTTP endpoint (see #11599).
When we use the HTTP endpoint, we only distinguish between a JSON format and a raw string format, and this allows to handle any document structure (a notebook is a JSON document, a text file is a string document...).
But when saving from the back-end, we handle `y-py` data structures: a Map, an Array or a Text object. A text file consists of a Text object for the whole content, a notebook consists of an Array for the cells, a Map for the metadata, and another Map for the state (from which we get the nbformat information). To compute the notebook source, we need to use these three objects.
This means that saving a document is document-specific, and this is an issue for extensions that handle new documents, because saving these documents will require a specific handling in the back-end. But the code responsible for saving the document in the back-end lives in the [Yjs handler](https://github.com/jupyterlab/jupyterlab/blob/a783add538f2fcd10ab4481fa5afd209c4d602c4/jupyterlab/handlers/yjs_echo_ws.py#L30-L36), which is not meant to be extensible currently.

### Proposed Solution

I think there are two options:
- make the Yjs handler extensible.
- require that a document has a unique top-level Y object (a Map), so that whatever the document we can compute its source by calling e.g. `top = ymap.to_json()`, and then get the source with `top["root"]`. For a text file, the content would be the value corresponding to the `root` key, for a notebook, the content would be the JSON under the `root` key.

cc @dmonad

**Comment by davidbrochart** [comment]
Another option that @SylvainCorlay mentioned, is having a declarative way of describing a Y document, e.g. for a notebook:
```
ydoc:
  cells: yarray
  meta: ymap
  state: ymap
format: json
    cells: cells.to_json()
    metadata: meta.to_json()["metadata"]
    nbformat: state.to_json()["nbformat"]
    nbformat_minor: state.to_json()["nbformatMinor"]
```
and for a text file:
```
ydoc:
  source: ytext
format: string
    source.to_string()
```
where the `ydoc` section describes the Y document structure, and the `format` section describes the stored document format, and how it is built from the `ydoc`. This kind of schema could then be used as a way to specify a shared document both in the back-end and in the front-end.

**Comment by davidbrochart** [comment]
Solved in #11599 with `YFile` and `YNotebook` in [ydoc.py](https://github.com/jupyterlab/jupyterlab/blob/82df0b635dae2c1a70a7c41fe7ee7af1c1caefb2/jupyterlab/handlers/ydoc.py). Other document structures can be added with the [jupyter_ydoc](https://github.com/jupyterlab/jupyterlab/blob/6c091561624110fe993f12fa0d9572c3ef418e26/setup.cfg#L85-L87) entry point.
